Introduce the `IntoIterator` trait and reimplement for-loops to use it

japaric opened the pull request:
As per [RFC #235][rfc], you can now do:

[rfc]: https://github.com/rust-lang/rfcs/blob/master/text/0235-collections-conventions.md#intoiterator-and-iterable

``` rust
let mut v = vec![1];

// iterate over immutable references
for x in &v {
    assert_eq!(x, &1);
}

// iterate over mutable references
for x in &mut v {
    assert_eq!(x, &mut 1);
}

// iterate over values, this consumes `v`
for x in v {
    assert_eq!(x, 1);
}
```

[breaking-change]s

For loops now "consume" (move) the iterator, this breaks iterating over mutable references to iterators, and also breaks multiple iterations over the same iterator:

``` rust
fn foo(mut it: &mut Iter) {  // `Iter` implements `Iterator`
    for x in it { .. }  //~ error: `&mut Iter` doesn't implement Iterator
}

fn bar() {
    for x in it { .. }  //~ note: `it` moved here
    for x in it { .. }  //~ error: `it` has been moved
}
```

Both cases can be fixed using the `by_ref()` adapter to create an iterator from the mutable reference:

``` rust
fn foo(mut it: &mut Iter) {
    for x in it.by_ref() { .. }
}

fn bar() {
    for x in it.by_ref() { .. }
    for x in it { .. }
}
```

This PR also makes iterator non-implicitly copyable, as this was source of subtle bugs in the libraries. You can still use `clone()` to explictly copy the iterator.

Finally, since the for loops are implemented in the frontend and use global paths to `IntoIterator`, `Iterator` and `Option` variants, users of the `core` crate will have to use add an `std` module to the root of their crate to be able to use for loops:

``` rust
#![no_std]

extern crate core;

fn main() {
    for x in 0..10 {}
}

#[doc(hidden)]
mod std {
    // these imports are needed to use for-loops
    pub use core::iter;
    pub use core::option;
}
```

---

r? @nikomatsakis @aturon 
cc #18424
closes #18045

rust-highfive commented:
r? @alexcrichton

(rust_highfive has picked a reviewer for you, use r? to override)

huonw commented:
Can we do this without expanding in the front end?

japaric commented:
@huonw @nikomatsakis was interested in an implementation that used the frontend (hence this prototype)

huonw commented:
Also, expanding to include `_ => break` will mean that refutable patterns like `for 1 in iter` will "work", which seems like something that we don't want to allow.

japaric commented:
(I think @aturon and @nikomatsakis mainly want to measure the size of the fallout and the change in ergonomics of these new for-loop semantics before commiting to it. This implementation is minimal work to measure that. We can come up with a better implementation if we decide to actually do this.)

alexcrichton commented on src/libcollections/str.rs:
```diff
@@ -273,7 +273,7 @@ impl<'a> Iterator for Recompositions<'a> {
             match self.state {
                 Composing => {
-                    for ch in self.iter {
+                    while let Some(ch) = self.iter.next() {
```
Could some loops like this be altered to:

```rust
for ch in self.iter.by_ref() { /* ... */ }
```

That may avoid consuming the entire iterator.

alexcrichton replied:
Actually, if it helps in basically all cases, I wonder if we could spec it like this...

japaric replied:
I had forgotten about `by_ref()`! I think we can use that instead of most (all?) the `while let`s.

> Actually, if it helps in basically all cases, I wonder if we could spec it like this...

You mean using `by_ref()` in the for loop expansion?

japaric commented:
Updated to use the `by_ref()` adapter instead of `while let`.

nikomatsakis commented:
Some thoughts:

(1) As I mentioned to @japaric on IRC, I think we can expand to `match <iterator> { mut iter => ... }` instead of `{ let mut iter = <iterator>; .. }` to resolve the matter of temporary lifetimes.

(2) Personally, I prefer to do this expansion than to have for loops in the AST that typeck/borrowck/everybody-else has to deal with. Basically desugaring makes the rest of the compiler simpler. However, the catch would be if it negatively affects error messages and so forth. So I'm curious @japaric as to the impact there. There is also the problem of our inability to really name things in std reliably, though this is ameliorated by the fact that `#![no_std]` is not stable yet.

jroesch commented:
@japaric and @nikomatsakis As a fan of desugaring as much as possible I think this seems like a good idea. If error messages become a problem you could use a similar scheme to what they do with Pyret where you provide a tag (either in some table) or directly on the AST for which rewrite rule you used, and then use that information when reporting errors by re-sugaring it then pretty printing.

japaric commented:
Updated PR with @nikomatsakis suggestions (`match IntoIterator` + UFCS with absolute paths, see expansion in the top comment), which got rid of the temporary lifetime issue, reduced the size of the import fallout and removed the need to have `IntoIterator` in the prelude. I'm now blocked on a type inference bug, but @nikomatsakis is aware of it.

I've collected some error messages from 3 cfail tests related to for-loops:

- for-loop-bogosity.rs

``` rust
struct MyStruct {
    x: isize,
    y: isize,
}

impl MyStruct {
    fn next(&mut self) -> Option<isize> {
        Option::Some(self.x)
    }
}

fn main() {
    let mut bogus = MyStruct {
        x: 1,
        y: 2,
    };
    for x in bogus {
//~^  old-ERROR has type `MyStruct` which does not implement the `Iterator` trait
//~^^ new-ERROR the trait `iter::Iterator` is not implemented for the type `MyStruct`
    }
}
```

The error message changed its wording but the meaning is the same.

- for-loop-refutable-pattern-error-message.rs
``` rust
fn main() {
    for
        &1i8
//~^  old-ERROR refutable pattern in `for` loop binding
//~^^ new-ERROR non-exhaustive patterns: `Some(&_)` not covered
        in [1i8].iter() {}
}
```

The message is more obscure now.

- for-loop-type-error.rs

``` rust
fn main() {
    let x = () + (); //~ ERROR binary operation

    // this shouldn't have a flow-on error:
    // japaric: and it doesn't with the either version of for loops
    for _ in x {}
}
```

Behavior remains the same.

alexcrichton commented on src/libcore/lib.rs:
```diff
@@ -65,6 +65,8 @@
 #![feature(on_unimplemented)]
 #![deny(missing_docs)]
+// SNAP 9e4e524 remove `allow(unused_mut)` after a snapshot
+#![allow(unused_mut)]
```
Now that we have `cfg_attr`, could this be `#![cfg_attr(stage0, allow(unused_mut))]`? Anything with the string "stage0" in it is a pretty easy flag for something to clean out when dealing with snapshots.

nikomatsakis commented:
@japaric and I were talking on IRC. My feeling is that the error in this example:

```rust
fn main() {
    for
        &1i8
//~^  old-ERROR refutable pattern in `for` loop binding
//~^^ new-ERROR non-exhaustive patterns: `Some(&_)` not covered
        in [1i8].iter() {}
}
```

is actually better. I doubt most "lay people" know what a "refutable pattern" is. (I personally have to work out from first principles which is "irrefutable" and which is "refutable".) But having a concrete counter example seems immediately understandable. It might be better still to combine the two messages:

    "refutable pattern in `for` loop binding: `Some(&_)` not covered"

huonw commented:
@nikomatsakis I think introducing the `Some` out of nowhere is confusing (and is not very useful since it happens for all invalid iterators, and is not related to how one might fix it, in the common case).

japaric commented:
This PR is ready! (passed `check-stage1`, running full `make check` now)

re: error message on refutable patterns, I decide to reuse the old message/diagnostic code, we can customize it later if desired.

japaric commented:
rebased and added tests for issues that new for loops fix. Also fixed a recursive call that the new `unconditional_recursion` lint pointed out. (last 3 commits are new)

japaric commented:
Just a heads up, until #21637 gets fixed, we won't be able to use the `IntoIterator` trait to build iterator-centric APIs. Code like ICEs right now:

``` rust
fn concat<I: IntoIterator>(it: I) -> String where <I::Iter as Iterator>::Item: Str {
    unimplemented!();
}
```

alexcrichton commented on src/failures:
```diff
@@ -0,0 +1,7 @@
+std
```
oops!

japaric replied:
ugh..
/me snipes file

eddyb commented:
Do we need `by_ref`? Can't we just use `&mut it` to achieve the same effect? Seems more consistent with iterating over vectors.

japaric commented:
rebased and reverted workarounds for #21245 because the fix landed.

@eddyb we could `impl<I: ?Sized + Iterator> Iterator for &mut I`, and maybe even remove the `by_ref()` adapter (?). One thing that makes me slightly uncomfortable with that change is that `it.count()` and `(&mut it).count()` would have different semantics, when in general `foo.method()` and `(&mut foo).method()` are equivalent when `foo` takes `&mut self`.

eddyb commented:
Neither of those will leave behind an unconsumed version of the iterator, it just acts as a manual reborrow, avoiding moves (if `count` takes `self` by value, that is. I don't see the difference if it takes a mutable ref).

Ericson2314 commented:
I mentioned on IRC once that it would be nice to allow using the iterator in the body of the for loop. But it seems this is not possible given `IntoIter` without a too-clever desugar? I guess `while-let` isn't too bad.

Also, does this mean the "curious inner module" is needed for for loops with `#![no_std]` ?

P1start commented:
@Ericson2314 Yes, the ‘curious inner module’ is needed, as noted near the end of the PR description.

Using the iterator in the for loop is #8372.

Ericson2314 commented:
@P1start Thank you. Somehow, I saw niko's comment, yet missed the mention at the top.

nikomatsakis commented:
@bors r+ 60abb3b

japaric commented:
@bors: r=nikomatsakis b9a9030

japaric commented:
I had to add a `#![feature(core)]` to some cfail tests to make them pass, which seem weird because AFAIK unmatched *warnings* don't make cfail tests fail. I'll investigate.

oli-obk commented:
@japaric: unmatched warnings fail cfail tests: https://github.com/rust-lang/rust/blob/master/src/compiletest/runtest.rs#L1011 only 'help' and others are ignored.

japaric commented:
@oli-obk Oh, I see. Thanks! This is my first time seeing a cfail test fail due to an unmatched warning, and since I've hardly seen a `//~ warning` annotation, I was under the impression that they were optional like `//~ help`, but then I guess most cfail tests don't throw warnings because they error before the lint pass!

bors commented:
:hourglass: Testing commit b9a9030 with merge 56b0e32...

bors commented:
:broken_heart: Test failed - [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/3032)

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit b9a9030 with merge e47284b...

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit b9a9030 with merge e513946...

bors commented:
:broken_heart: Test failed - [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/3372)

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit b9a9030 with merge 581a5c8...

bors commented:
:broken_heart: Test failed - [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/3353)

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit b9a9030 with merge 692c620...

bors commented:
:broken_heart: Test failed - [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/3373)

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit b9a9030 with merge 72eab3e...

taralx commented:
If I'm reading this right, for loops cannot be use in std-free code then? What happened to core-only?

taralx commented:
The RFC didn't mention this "mod std" business, and I'm a bit uncomfortable with it. Why not directly access the core crate?

japaric commented:
> Why not directly access the core crate?

If you mean expanding to `::core::iter::IntoIterator`, that won't work in most cases, since most crates are linked to `std` but not to `core` (they have access to `core` stuff because of re-exports).

> The RFC didn't mention this "mod std" business, and I'm a bit uncomfortable with it.

You already have to use the `mod std` to use the `#[derive]` syntax extension and/or slicing syntax in `#![no_std]` crates. At the moment this required for stuff that's implemented in the frontend, but we'd definitely like to improve this situation in the future.

japaric commented:
@taralx see #21912 ;-)

alexchandel commented:
The [relevant section](http://doc.rust-lang.org/reference.html#for-expressions) of the Rust Reference ought to be updated to reflect this, cc @steveklabnik
